qiskit/aqua/operators/state_fns/operator_state_fn.py

```diff
@@ -0,0 +1,210 @@
+            raise ValueError(
+                'to_matrix will return an exponentially large matrix,'
+                ' in this case {0}x{0} elements.'
+                ' Set massive=True if you want to proceed.'.format(2 ** self.num_qubits))
+
+        return self.primitive.to_matrix() * self.coeff
+
+    def to_matrix_op(self, massive: bool = False) -> OperatorBase:
+        """ Return a MatrixOp for this operator. """
+        return OperatorStateFn(self.primitive.to_matrix_op(massive=massive) * self.coeff,
+                               is_measurement=self.is_measurement)
+
+    def to_matrix(self, massive: bool = False) -> np.ndarray:
+        r"""
+        Note: this does not return a density matrix, it returns a classical matrix
+        containing the quantum or classical vector representing the evaluation of the state
+        function on each binary basis state. Do not assume this is is a normalized quantum or
+        classical probability vector. If we allowed this to return a density matrix,
+        then we would need to change the definition of composition to be ~Op @ StateFn @ Op for
+        those cases, whereas by this methodology we can ensure that composition always means Op
+        @ StateFn.
+
+        Return numpy vector of state vector, warn if more than 16 qubits to force the user to set
+        massive=True if they want such a large vector.
+
+        Args:
+            massive: Whether to allow large conversions, e.g. creating a matrix representing
+                over 16 qubits.
+
+        Returns:
+            np.ndarray: Vector of state vector
+
+        Raises:
+            ValueError: Invalid parameters.
+        """
+
+        if self.num_qubits > 16 and not massive:
+            raise ValueError(
+                'to_vector will return an exponentially large vector, in this case {0} elements.'
+                ' Set massive=True if you want to proceed.'.format(2 ** self.num_qubits))
+
+        # Operator - return diagonal (real values, not complex),
+        # not rank 1 decomposition (statevector)!
+        mat = self.primitive.to_matrix()
+        # TODO change to weighted sum of eigenvectors' StateFns?
+
+        # ListOp primitives can return lists of matrices (or trees for nested ListOps),
+        # so we need to recurse over the
+        # possible tree.
+        def diag_over_tree(t):
+            if isinstance(t, list):
+                return [diag_over_tree(o) for o in t]
+            else:
+                vec = np.diag(t) * self.coeff
+                # Reshape for measurements so np.dot still works for composition.
+                return vec if not self.is_measurement else vec.reshape(1, -1)
+
+        return diag_over_tree(mat)
+
+    def to_circuit_op(self) -> OperatorBase:
+        r""" Return ``StateFnCircuit`` corresponding to this StateFn. Ignore for now because this is
+        undefined. TODO maybe diagonalize here."""
+        return self
```

Comment:
Given the comment, is returning self the "right" thing to do or should it raise a not implemented error?

Reply:
Good call, fixed